### ISSUE TITLE
sf.kde flips raster output

### ISSUE THREAD

**Issue by d-pflugmacher** [jeffreyevans/spatialEco#50]
Dear Jeff, 

The sf.kde() functions gives me a raster output with a different orientation. The output does not match the locations of the input vector features. Here is my code.

```{r}
library(spatstats)
library(terra)

data("nbfires")

# convert ppp to sf
nbfires <- rescale(nbfires)
nbfires <- subset(nbfires, fnl.size > 0)
boundary <- st_as_sf(nbfires$window)
df.fires <- data.frame(x=nbfires$x, y=nbfires$y, marks(nbfires))
fires <- st_as_sf(df.fires, coords=c("x", "y"))

fire_count <- terra::rasterize(fires, ras, fun = "length", background = 0)
plot(fire_count)

fire_kde <- spatialEco::sf.kde(fires, bw = c(50, 50), ref = fire_count, standardize = T)

# wrong orientation
plot(fire_kde)
plot(boundary, add=T)

# right orientation - but does not match
plot(flip(flip(t(fire_kde), 'horizontal'), 'vertical'))
plot(boundary, add=T)
```

Thank you!
Dirk

**Comment by jeffreyevans** [comment]
Dirk,
I am sorry for the inconvenience. In the latest release of spatialEco, I gutted this function and switched the kernel density estimate model from MASS to ks. As such I am now using some of the internal array functions in the ks package and have not tested the function intensively.

I am getting some odd behavior where this happens on some data and not on other and though that I had fixed it. I am wondering if the CSR and global location of the data are causing issues with ks that I had not addressed. Would it be possible for you to share some of the data you are having issues with so I can recreate this behavior? If you are concerned about propriety data you can send a fuzzified version. I really only need [x,y] coordinates and the associated csr as, I can create dummy weights for error checking the weighted version.

Best,
Jeff

Jeffrey S. Evans, Ph.D., | Senior Landscape Ecologist & Biometrician
The Nature Conservancy | Global Protect, Science
Visiting Professor | University of Wyoming | Zoology & Physiology
Laramie, WY | ***@***.******@***.***> | (970) 672-6766<tel:(970)%20672-6766>
ORCID-ID 0000-0002-5533-7044
TNC Biography<https://www.nature.org/en-us/about-us/who-we-are/our-people/jeff-evans/> – Google Scholar<https://scholar.google.com/citations?user=FwU_ppgAAAAJ&hl=en>

From: Dirk Pflugmacher ***@***.***>
Sent: Friday, December 2, 2022 2:22 PM
To: jeffreyevans/spatialEco ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [jeffreyevans/spatialEco] sf.kde flips raster output (Issue #50)


Dear Jeff,

The sf.kde() functions gives me a raster output with a different orientation. The output does not match the locations of the input vector features. Here is my code.

library(spatstats)

library(terra)



data("nbfires")



# convert ppp to sf

nbfires <- rescale(nbfires)

nbfires <- subset(nbfires, fnl.size > 0)

boundary <- st_as_sf(nbfires$window)

df.fires <- data.frame(x=nbfires$x, y=nbfires$y, marks(nbfires))

fires <- st_as_sf(df.fires, coords=c("x", "y"))



fire_count <- terra::rasterize(fires, ras, fun = "length", background = 0)

plot(fire_count)



fire_kde <- spatialEco::sf.kde(fires, bw = c(50, 50), ref = fire_count, standardize = T)



# wrong orientation

plot(fire_kde)

plot(boundary, add=T)



# right orientation - but does not match

plot(flip(flip(t(fire_kde), 'horizontal'), 'vertical'))

plot(boundary, add=T)

Thank you!
Dirk

**Comment by d-pflugmacher** [comment]
Jeff,

no worries. I've encountered the issue with the `nbfires` dataset built-in the package `spatstat`. I updated my code below, so you can test it on your computer. The dataset does not have a CRS. However, I also just replicated the same behaviour with another dataset (attached). Hope this helps.

terra 1.6.47
sf 1.0-9
spatialEco 2.0-0 from CRAN

```{r}
library(terra)
library(spatstat)
library(sf)

data("nbfires")
class(nbfires)

# convert ppp to sf
nbfires <- spatstat.geom::rescale(nbfires)
boundary <- st_as_sf(nbfires$window)
df.fires <- data.frame(x=nbfires$x, y=nbfires$y, marks(nbfires))
fires <- st_as_sf(df.fires, coords=c("x", "y"))

ras <- rast(ext(fires))
res(ras) <- c(10, 10)
fire_kde <- spatialEco::sf.kde(fires, bw = c(50, 50), ref = ras, standardize = T)

# wrong orientation
plot(fire_kde)
plot(boundary, add=T)

# right orientation - but does not match
plot(flip(flip(t(fire_kde), 'horizontal'), 'vertical'))
plot(boundary, add=T)
```

Best,
Dirk



[abiesalba_carp.gpkg.zip](https://github.com/jeffreyevans/spatialEco/files/10145144/abiesalba_carp.gpkg.zip)

**Comment by jeffreyevans** [comment]
Dirk,
Sorry for not getting to this sooner. I did find a bug in the sf.kde function that was flipping and distorting the data. This has been fixed in the development version 2.0-01, currently on GutHub and soon will be on [R-Universe](https://jeffreyevans.r-universe.dev/ui#package:spatialEco). I will push a patch to CRAN in a week or two. 

However, there was an additional issue with the spatstat nbfires data being in a stereographic projection (EPSG:2953) with a rotated extent. Admittedly, I was a bit flummoxed by this one because, I addressed the original issue and a rotation issues persisted. The bug and the projection is why you were having to transpose then flip horizontally and vertically. Please test and let me know if any additional issues arise. You can install the development version using: `remotes::install_github("jeffreyevans/spatialEco", force=TRUE)`

With the new version of the function, here is a workflow that yields, hopefully, correct results. 

```
library(terra)
library(spatstat)
library(sf)
library(spatialEco)

data(nbfires)
nbfires <- spatstat.geom::rescale(nbfires)

boundary <- st_as_sf(nbfires$window)
  st_crs(boundary) <- sf::st_crs('EPSG:2953')

df.fires <- data.frame(x=nbfires$x, y=nbfires$y, marks(nbfires))
  fires <- st_as_sf(df.fires, coords=c("x", "y"))
    st_crs(fires) <- sf::st_crs('EPSG:2953')

fires <- st_transform(fires, st_crs('ESRI:102001'))
  # fires <- remove_duplicates(fires)
boundary <- st_transform(boundary, st_crs('ESRI:102001'))  

ras <- rast(ext(fires))
  res(ras) <- c(10, 10)

fire_kde <- sf.kde(fires, bw = c(50, 50), 
            ref = ras, standardize = TRUE)

plot(fire_kde)
  plot(boundary, add=TRUE)
  plot(st_geometry(fires), pch=20, cex=0.25, add=TRUE)
```

Best,
Jeff
